### ISSUE TITLE
How to type fast with launcher mode. Not sure how to adjust parameters so it works like in old Karabiner

### ISSUE THREAD

**Issue by nikitavoloboev** [tekezo/Karabiner-Elements#1484]
I have an issue with launcher mode rules. For example I have [this rule in Elements](https://github.com/nikitavoloboev/dotfiles/blob/master/karabiner/karabiner.json#L1852) which turns w key into a launcher mode key.

However the issue is that if I type quickly wa for example, only a will be inserted into the screen. Also I make every key on my keyboard into a launcher key and would love to make it work just like in old Karabiner where I could write fast and without any issues.

[Here](https://github.com/nikitavoloboev/dotfiles/blob/master/karabiner/karabiner.json#L9) are the parameters I have set now. I tried to tweak them but nothing seems to work. I am not sure what values should these parameters be to make launcher mode work like in old Karabiner.

@tekezo Is this feature still not supported in Elements?  

Thank you a lot for the help.

**Comment by apm1467** [comment]
If you want to use normal letter keys as mod trigger keys, you will need the keyword `simultaneous`. 

You can take a look at this [Simple Vi Mode](https://github.com/pqrs-org/KE-complex_modifications/blob/master/docs/json/personal_tekezo_simple_vi_mode.json) example. Note how `simultaneous `, `key_down_order` and `key_up_order ` are set. With them set you will be able to type fast without false trigger.

**Comment by nikitavoloboev** [comment]
> You can take a look at this Simple Vi Mode example

I just took a look at it and am a bit confused. This Vi Mode simply turns h j k l into arrow keys. I would like instead for it to trigger arrow keys on hjkl if s key is pressed. I could do that in old Karabiner.

**Comment by nikitavoloboev** [comment]
Reading through the code it seems that hjkl should only be triggered on vim mode, but for me if I simply press h key without any keys, it does left arrow.

**Comment by apm1467** [comment]
This rule turns h j k l into arrow keys only when the s key is hold down. If s is not hold down, they should be just normal h j k l keys. 

Doesn't it work right for you? Are you running the latest version of Karabiner-Elements? You can import this rule directly from [here](https://pqrs.org/osx/karabiner/complex_modifications/#personal_tekezo_simple_vi_mode) if you'd like to test.

**Comment by nikitavoloboev** [comment]
> Are you running the latest version of Karabiner-Elements?

I am running 11.6.0 because this is last version that works on 10.11.6. I rely on my old Karabiner configuration too much to upgrade fully to Sierra.

Can you perhaps give a hand at helping me convert my configuration? I would be forever grateful to you for that. I've been trying to do it for some time now but I can't do it nicely. This is my old [Karabiner config](https://github.com/nikitavoloboev/dotfiles/blob/645eb48eb78a12d006eb31c57f3587a9a08f81dc/karabiner/private.xml).

It seems that in Elements this will be over 100,000 lines long. Can Elements even handle a file size this big?

**Comment by apm1467** [comment]
Back in version 11.6.0, the `simultaneous` function hadn't been implemented, so there's no good way to achieve what you want.

Your config file looks pretty structural. You can consider writing a small script to generate the final JSON config file. [The source code](https://github.com/pqrs-org/KE-complex_modifications/blob/master/src/json/personal_tekezo_simple_vi_mode.json.rb) of the above Simple Vi Mode, for example, is a `.rb` ruby file. When we run `$ make` in terminal, it will generate the final 345 lines of JSON file from only 82 lines of ruby code. 

The chance is good that you can write a simple function that generates JSON config rules, and you just call it multiple times with different arguments, and it will generate thousands of lines of JSON.

**Comment by nikitavoloboev** [comment]
> When we run $ make in terminal

There are many other `.rb` files in the src/json folder. Where do I run make?

Also I am curious how it would work, does it just generate JSON and prepend it to karabiner.json? What if I want to change something to the .rb file, when I run it again, will it swap the generated rule or just add another rule to the previously generated one?

**Comment by nikitavoloboev** [comment]
Oh okay I got it. I ran `ruby personal_tekezo_simple_vi_mode.json.rb | pbcopy` and then pasted the rule in the karabiner.json.

Is it possible to do it like this though?

1. Edit the ruby code script
2. Run it in some way and it in smart way swaps the rule it generated previously with what is newly generated?

**Comment by apm1467** [comment]
Other than karabiner.json, there are single file complex modification rules stored at your `~/.config/karabiner/assets/complex_modifications`. When you import complex modification rules from the internet, they are all stored there.

So you can do something like:

```
$ ruby personal_tekezo_simple_vi_mode.json.rb > ~/.config/karabiner/assets/complex_modifications/simple_vi_mode.json
```

And you will see this rule directly in Karabiner-Elements > Complex Modifications > Add rule. 

Every time you run the above command, the JSON file there named simple_vi_mode.json will be overwritten with the new result. But you have to manually click "Remove" then "Add rule" in Karabiner-Elements again for the change to take effect.

**Comment by nikitavoloboev** [comment]
> But you have to manually click "Remove" then "Add rule" in Karabiner-Elements again for the change to take effect.

Is it possible to automate/skip this step? I plan to make changes to the ruby script often and currently plan generate ALL of the JSON [from one script](https://github.com/nikitavoloboev/karabiner-generator/blob/master/karabiner-generate.rb) and all the rules from it. 

So I want to make changes to the script, press save and on save it will generate JSON and transfer that to Elements.

I can run a script on save with Hazel no problem but I don't want to go through the manual process of 

> you have to manually click "Remove" then "Add rule" in Karabiner-Elements again for the change to take effect.

**Comment by apm1467** [comment]
You may create something like a Keyboard Maestro micro to let the mouse click these buttons automatically.

**Comment by nikitavoloboev** [comment]
> create something like a Keyboard Maestro micro

Will do that for now, yeah.

@tekezo Would it be possible to add support for automatically reloading rules in Karabiner?

Thank you.

**Comment by Jimmy34742** [comment]
@nikitavoloboev I'm a touch typist. I worked hard to adjust to my mode launcher. It forced me to eliminate all overlapping of keystrokes, which improved my typing, but was ultimately quite speed-limited. After allowing a long time to get used to both, there was no question that my simultaneous launcher worked better for me, and I only rarely run into typing timing glitches, and only then when I don't have my keyboard ergonomically positioned. I found that combinations of 4 or more keys slows down my typing too much because the unnaturalness of it makes me set before I type. 2-key combos work the best, and my present launcher is 76 2-key combos and 30 3-keys. I also find it much easier to memorize a large number of randomly assigned 2-key codes (mnemonical ones are too few), probably from Unix command line days. 

I know that you have a lot more key combinations than that, so this is just feedback for comparison. Perhaps if you cleverly re-partitioned your UI into simultaneous and modes that might work. But I found that any key used to enter a mode became a major bottleneck for typing, no matter how much I tried to make it work.
